README.md
# OME Common C++

OME Common is a standalone C++ library required by other OME C++
projects for common functionality which is not readily available from
the C++ Standard Library.  This includes basic portability functions,
to wrapping other libraries to make them usable with Modern C++
programming practices.

It serves a similar purpose to the OME formats-common Java library,
with some shared functionality, though for the most part they are
quite different.

Links
-----

- [Documentation](https://docs.openmicroscopy.org/latest/ome-files-cpp/)
- [API reference](https://docs.openmicroscopy.org/latest/ome-files-cpp/ome-common/api/html/namespaces.html)

Purpose
-------

OME Common's primary purpose is to ensure that a certain level of
basic functionality is provided for all platforms and compilers
supported by OME C++ projects.  It currently includes:

- Missing C++ standard library functionality:

  * regex

- Missing Boost functionality:

  * Endian conversions
  * Filesystem absolute path determination
  * Variant and MPL set up
  * Units (extended datatypes for all units used by the OME data
    model)

- Extra functionality:

  * Boolean type for iterable 8-bit mask pixel data
  * logging
  * installation path determination
  * memory streams
  * string portability and helper functions
  * XML parsing (Xerces-C++ Modern C++ RAII wrappers)


More information
----------------


For more information, see the [documentation](https://docs.openmicroscopy.org/latest/ome-files-cpp/).


Pull request testing
--------------------

We welcome pull requests from anyone, but ask that you please verify the
following before submitting a pull request:

 * verify that the branch merges cleanly into ```master```
 * verify that the branch compiles
 * run the unit tests (```ctest -V```) and correct any failures
 * make sure that your commits contain the correct authorship information and,
   if necessary, a signed-off-by line
 * make sure that the commit messages or pull request comment contains
   sufficient information for the reviewer(s) to understand what problem was
   fixed and how to test it


NEWS.md
5.5.1 (unreleased)
------------------

* No changes

5.5.0 (2017-11-28)
------------------

* Use CMP0067 to enable standard setting in feature tests ([#47](https://github.com/ome/ome-common-cpp/pull/47))
* CMake 3.4 is the minimum version ([#48](https://github.com/ome/ome-common-cpp/pull/48))
* Use C++11 `<thread>` and `<mutex>` ([#46](https://github.com/ome/ome-common-cpp/pull/46))
* Restore support for Boost 1.53 ([#49](https://github.com/ome/ome-common-cpp/pull/49))
* Add support for Boost 1.65 and 1.65.1 ([#50](https://github.com/ome/ome-common-cpp/pull/50))
* Conditionally add `Boost::log` and `Boost::log_setup` to exported configuration ([#51](https://github.com/ome/ome-common-cpp/pull/51))
* Drop embedded copies of CMake `FindBoost`, `FindXalanC` and `FindXercesC`
  ([#53](https://github.com/ome/ome-common-cpp/pull/53))

5.4.2 (2017-06-12)
------------------

* Doxygen improvements ([#38](https://github.com/ome/ome-common-cpp/pull/38), [#39](https://github.com/ome/ome-common-cpp/pull/39))
* CMake: use C++ standard variables as documented ([#40](https://github.com/ome/ome-common-cpp/pull/40))

5.4.1 (2017-05-23)
------------------

* Add documentation links ([#35](https://github.com/ome/ome-common-cpp/pull/35))
* Add support for Boost 1.64 ([#37](https://github.com/ome/ome-common-cpp/pull/37))

5.4.0 (2017-02-09)
------------------

* C++11 is now the minimum required language version, with C++14 being used when available ([#29](https://github.com/ome/ome-common-cpp/pull/29))
* enabled the use of a restricted number of C++11 features, including enum class, nullptr, initializer lists, range-based for loops and type traits ([#31](https://github.com/ome/ome-common-cpp/pull/31))
* enabled the use of C++11 syntax changes including `<::` not being interpreted as a trigraph and `»` being used to close nested templates instead of `> >` ([#31](https://github.com/ome/ome-common-cpp/pull/31))
* Google Test (gtest) is no longer built separately in each source component; the latest gtest release now allows use as a conventional library ([#28](https://github.com/ome/ome-common-cpp/pull/28))
* source releases are now made directly from git with `git archive` ([#32](https://github.com/ome/ome-common-cpp/pull/32))
* Add support for Boost 1.63 ([#30](https://github.com/ome/ome-common-cpp/pull/30))
* Boost MPL size limits are now set with MSVC only; they are no longer required for other compilers with C++11
* `<ome/compat/array.h>`, `<ome/compat/cstdint.h>`, `<ome/compat/memory.h>` and `<ome/compat/tuple.h>` are now deprecated in favor of the standard headers; they wrap the standard headers for backward compatibility but will be removed in the next breaking release ([#31](https://github.com/ome/ome-common-cpp/pull/31))

5.3.2 (2016-10-12)
------------------

* Boost 1.54 is the minimum supported version ([#25](https://github.com/ome/ome-common-cpp/pull/25))
* Add support for Boost 1.62 ([#26](https://github.com/ome/ome-common-cpp/pull/26))

5.3.1 (2016-09-18)
------------------

* Document all namespaces in API reference ([#24](https://github.com/ome/ome-common-cpp/pull/24))

5.3.0 (2016-07-29)
------------------

* Added Base64 conversion functions ([#16](https://github.com/ome/ome-common-cpp/pull/16))
* Corrected Xerces UTF8 to UTF-16 transcoding for empty strings ([#18](https://github.com/ome/ome-common-cpp/pull/18))
* Added rankine temperature conversions ([#17](https://github.com/ome/ome-common-cpp/pull/17))
* Added MSVC workarounds for system macros with names matching unit names ([#22](https://github.com/ome/ome-common-cpp/pull/22))
* Removed VS2012 workarounds ([#15](https://github.com/ome/ome-common-cpp/pull/15))
* Improved CMake configuration export ([#21](https://github.com/ome/ome-common-cpp/pull/21))
* Added BSD licence text ([#23](https://github.com/ome/ome-common-cpp/pull/23))

5.2.0 (2016-04-23)
------------------

* Decouple ome-common-cpp component from Bio-Formats ([#1](https://github.com/ome/ome-common-cpp/pull/1), [#9](https://github.com/ome/ome-common-cpp/pull/9))
* Use external gtest ([#2](https://github.com/ome/ome-common-cpp/pull/2))
* Add support for Boost 1.60 ([#4](https://github.com/ome/ome-common-cpp/pull/4))
* Add Xalan-C support ([#7](https://github.com/ome/ome-common-cpp/pull/7))
* Rename all uses of bioformats to ome-common ([#10](https://github.com/ome/ome-common-cpp/pull/10))
* Add `source-archive` script to generate a source release ([#12](https://github.com/ome/ome-common-cpp/pull/12), [#13](https://github.com/ome/ome-common-cpp/pull/13), [#14](https://github.com/ome/ome-common-cpp/pull/14))


lib/ome/common/base64.h
/*
 * #%L
 * OME-COMMON C++ library for C++ compatibility/portability
 * %%
 * Copyright © 2016 Open Microscopy Environment:
 *   - Massachusetts Institute of Technology
 *   - National Institutes of Health
 *   - University of Dundee
 *   - Board of Regents of the University of Wisconsin-Madison
 *   - Glencoe Software, Inc.
 * %%
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * The views and conclusions contained in the software and documentation are
 * those of the authors and should not be interpreted as representing official
 * policies, either expressed or implied, of any organization.
 * #L%
 */

#ifndef OME_COMMON_BASE64_H
#define OME_COMMON_BASE64_H

#include <cctype>
#include <cstdint>
#include <iterator>
#include <stdexcept>
#include <string>

namespace ome
{
  namespace common
  {

    namespace detail
    {

      /**
       * Append a Base64-encoded character to a Base64-encoded string.
       *
       * Line breaks will be automatically appended if required.
       *
       * @param encoded the Base64-encoded string.
       * @param value the value to append.
       * @param chars the number of Base64 characters written (used
       * for linebreaking).
       * @param linebreak the position at which to break a line; zero
       * to disable line breaks.
       */
      inline void
      output_base64_char(std::string& encoded,
                         uint8_t      value,
                         uint64_t     chars,
                         uint8_t      linebreak)
      {
        // Static mapping using simple offsets.
        static const char * const base64_chars =
          "ABCDEFGHIJKLMNOPQRSTUVWXYZ" //  0-25
          "abcdefghijklmnopqrstuvwxyz" // 26-51
          "0123456789+/=";             // 52-63

        encoded += base64_chars[value];
        if (linebreak && (chars + 1) % linebreak == 0)
          encoded += '\n';
      }

      /**
       * Get the value corresponding to a Base64-encoded character.
       *
       * The returned value will be in the range 0-63 for valid
       * Base64 characters.  98 is used for padding.
       *
       * @param c the character to look up.
       * @returns the value.
       * @throws std::runtime_error if an invalid character is used.
       */
      inline uint8_t
      base64_value(char c)
      {
        // Static mapping; due to the nature of the encoding, it's sparse.
        // Without C0 (-0x20) to reduce size. 98=pad, 99=invalid.
        static const uint8_t base64_codes[] =
          {
            // ' ' ! " # $ % & ' ( ) * + , - . /
            99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 62, 99, 99, 99, 63,
            // 0 1 2 3 4 5 6 7 8 9 : ; < = > ?
            52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 99, 99, 99, 98, 99, 99,
            // @ A B C D E F G H I J K L M N O
            99, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14,
            // P Q R S T U V W X Y Z [ \ ] ^ _
            15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 99, 99, 99, 99, 99,
            // ` a b c d e f g h i j k l m n o
            99, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40,
            // p q r s t u v w x y z
            41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51
          };

        if (c < 0x20 || c > 'z')
          throw std::runtime_error("Invalid Base64 input: character outside permitted range");
        uint8_t v = base64_codes[c - 0x20];
        if (v == 99)
          throw std::runtime_error("Invalid Base64 input: character outside permitted range");
        return v;
      }

      /**
       * Get the next character value from a Base64-encoded string.
       *
       * Any whitespace in the string will be automatically skipped.
       *
       * The returned value will be in the range 0-63 for valid Base64
       * characters.  98 is used for padding. 255 is used if the end
       * of the string was reached (used to detect end of valid input
       * in the presence of trailing whitespace).
       *
       * @param pos the current position in the string.
       * @param end the end of the string.
       * @returns the character value, padding or end of string marker.
       * @throws std::runtime_error if an invalid character is encountered.
       */
      inline uint8_t
      next_base64_value(std::string::const_iterator& pos,
                        std::string::const_iterator  end)
      {
        while (pos != end)
          {
            if (!std::isspace(*pos))
              {
                uint8_t value = base64_value(*pos++);
                return value;
              }
            ++pos;
          }
        return 255;
      }

    }

    /**
     * Base64-encode a range of bytes.
     *
     * The iterators may be of any uint8_t iterator type, for example
     * a range in a container, raw pointers or any other compatible
     * type.
     *
     * @param begin the start of the byte range.
     * @param end the end of the byte range.
     * @param linebreak the position at which to break a line; zero
     * to disable line breaks.
     * @returns a Base64-encoded string.
     */
    template<typename Iterator>
    std::string
    base64_encode(Iterator begin,
                  Iterator end,
                  uint8_t  linebreak = 76)
    {
      std::string encoded;
      encoded.reserve((std::distance(begin, end) * 4) / 3);

      uint64_t bytes = 0U;
      uint64_t chars = 0U;
      uint8_t accum = 0U;
      for (Iterator i = begin; i != end; ++bytes)
        {
          uint8_t byte = *i; // Byte 1
          ++i;
          accum = byte >> 2;
          detail::output_base64_char(encoded, accum, chars++, linebreak); // Char 1
          accum = (byte & 0x3) << 4;
          if (i != end)
            {
              byte = *i; // Byte 2
              ++i;
              accum |= (byte & 0xF0) >> 4;
              detail::output_base64_char(encoded, accum, chars++, linebreak); // Char 2
              accum = (byte & 0x0F) << 2;
              if (i != end)
                {
                  byte = *i; // Byte 3
                  ++i;
                  accum |= (byte & 0xC0) >> 6;
                  detail::output_base64_char(encoded, accum, chars++, linebreak); // Char 3
                  accum = byte & 0x3F;
                  detail::output_base64_char(encoded, accum, chars++, linebreak); // Char 4
                }
              else
                {
                  detail::output_base64_char(encoded, accum, chars++, linebreak); // Char 3 (+pad)
                  encoded += '='; // Char 4 (pad)
                }
            }
          else
            {
              detail::output_base64_char(encoded, accum, chars++, linebreak); // Char 2 (+pad)
              encoded += "=="; // Chars 3 and 4
            }
        }

      return encoded;
    }

    /**
     * Decode a Base64-encoded string.
     *
     * Newlines and other whitespace breaking up the input are
     * permitted.
     *
     * @param base64 the Base64-encoded string.
     * @param dest the insert iterator for the destination
     * @throws std::runtime_error on invalid input.
     */
    template<typename InsertIterator>
    void
    base64_decode(std::string    base64,
                  InsertIterator dest)
    {
      bool pad_seen = false;
      uint8_t bytes[4];
      uint8_t output;
      for (std::string::const_iterator i = base64.begin();
           i != base64.end();)
        {
          // Get next 4 bytes.  If only whitespace remains, the first
          // byte will be 255.  Since the input is blocked into groups
          // of four characters, fetch four at once.
          for (int j = 0; j < 4; ++j)
            {
              bytes[j] = detail::next_base64_value(i, base64.end());
              if(j == 0 && bytes[j] == 255)
                break;
            }

          if (bytes[0] == 255) // End of input (expected)
            break;
          if (bytes[1] == 255 || bytes[2] == 255 || bytes[3] == 255) // End of input (unexpected)
            throw std::runtime_error("Invalid Base64 input: unexpected end of input");

          if (bytes[0] < 64 && bytes[1] < 64) // Valid input
            {
              if(pad_seen) // Padding only allowed at end.
                throw std::runtime_error("Invalid Base64 input: padding only permitted at end of input");

              output = bytes[0] << 2 | bytes[1] >> 4;
              *dest = output; // Byte 1
              if (bytes[2] < 64) // Skip if padded
                {
                  output = bytes[1] << 4 | bytes[2] >> 2;
                  *dest = output; // Byte 2
                  if (bytes[3] < 64) // Skip if padded
                    {
                      output = bytes[2] << 6 | bytes[3];
                      *dest = output; // Byte 3
                    }
                  else
                    {
                      pad_seen = true;
                    }
                }
              else
                {
                  pad_seen = true;
                }
            }
          else
            {
              throw(std::runtime_error("Invalid Base64 input: padding encountered unexpectedly"));
            }
        }
    }

    /**
     * Decode a Base64-encoded string into a container.
     *
     * Newlines and other whitespace breaking up the input are
     * permitted.
     *
     * @param base64 the Base64-encoded string.
     * @returns a container filled with the decoded bytes.
     * @throws std::runtime_error on invalid input.
     */
    template<typename Container>
    Container
    base64_decode(const std::string& base64)
    {
      Container decoded;

      decoded.reserve((base64.size() * 3) / 4);

      base64_decode(base64, std::back_inserter(decoded));

      return decoded;
    }

  }
}

#endif // OME_COMMON_BASE64_H

/*
 * Local Variables:
 * mode:C++
 * End:
 */
